### PR TITLE
Temporarily add authentication logging to help debug auth issues

### PR DESCRIPTION
**Pull request by tahb**
Recently our users (as well as others in MoJ) have been seeing lots of autherrors. Our temporary workaround has been to advise users to clear their cookies so we need to resolve this.

On request of the HAAR team we temporarily add in this debugging so that we can assist them debug Auth issues that we and other services have been experiencing. 

This approach is lifted from another team who have already added it[1] where we redefine the implementation of the callback `successReturnToOrRedirect` with our custom handling of the same but which lets us collect more information.

[1] https://github.com/ministryofjustice/hmpps-digital-prison-services/pull/84

Checked sign in still works locally in the happy case:

https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/912473/7b16bbad-61e6-46a6-ae80-c5d81b9b9748

# Release checklist

[Release process documentation](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/edit-v2/4247847062?draftShareId=a1c360ab-bd31-4db1-aae3-7cc002761de9).

## Pre-merge checklist

- [ ] Are any changes required to dependent services for this change to work?
  (eg. CAS API)
- [ ] Have they been released to production already?

## Post-merge checklist

- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to test
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to preprod
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to prod

### CODE REVIEW

File: server/middleware/setUpAuthentication.ts [hunk 1]
```diff
@@ -21,12 +22,35 @@
...
+        }
+        if (typeof returnTo === 'string' && returnTo.startsWith('/')) {
+          return res.redirect(returnTo)
+        }
+        return res.redirect('/')
+      })
+      return null
+    }
+    passport.authenticate('oauth2', authCallback)(req, res, next)
+  })
```
[the remark this answers is not in the document]
Comment (tahb) [reply server/middleware/setUpAuthentication.ts]: It's not visible but I've closed the alert with a message that I'm happy there's no expensive operations here that increase the risk of a DoS?

File: server/middleware/setUpAuthentication.ts [hunk 2]
```diff
@@ -21,12 +22,35 @@ export default function setUpAuth(): Router {
...
-      failureRedirect: '/autherror',
-    })(req, res, next),
-  )
+  router.get('/sign-in/callback', (req, res, next) => {
+    const authCallback: passport.AuthenticateCallback = (err, user, info) => {
+      if (err) {
+        return res.redirect('/autherror')
+      }
+      if (!user) {
+        const { message } = info as { message: string }
```
Comment (richpjames) [inline server/middleware/setUpAuthentication.ts]: Are we safe to assume the `message` will *always* be present in `info`? if not we're better off checking it exists before destructuring it

File: server/middleware/setUpAuthentication.ts [hunk 3]
```diff
@@ -21,12 +22,35 @@ export default function setUpAuth(): Router {
...
+        if (info && message === 'Unable to verify authorization request state.') {
+          // failure to due authorisation state not being there on return, so retry
+          logger.info('Retrying auth callback as no state found')
+          return res.redirect('/')
+        }
+        logger.info(`Auth failure due to ${JSON.stringify(info)}`)
+        return res.redirect('/autherror')
+      }
+      const { returnTo } = req.session
```
Comment (richpjames) [inline server/middleware/setUpAuthentication.ts]: Same here, if we're confident `returnTo` always exists in session then this is fine but if there's a chance it won't then we're better of checking it exists before destructuring.

**Review by richpjames — APPROVED** [review]
I've left a couple of comments but approving anyway as I assume this code has come from the HAAR team who know it to be safe

File: server/middleware/setUpAuthentication.ts (hunk 2 above)
Comment (tahb) [reply server/middleware/setUpAuthentication.ts]: I can't be sure so let's guard against it to be safe 👍

File: server/middleware/setUpAuthentication.ts (hunk 3 above)
Comment (tahb) [reply server/middleware/setUpAuthentication.ts]: I can't be sure so let's guard against it to be safe 👍

**Pull request merged**